src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -660,14 +659,16 @@ private static bool ShouldUseKnownLength(BoundCollectionExpression node, out int
             if (node is { Elements: [BoundCollectionExpressionSpreadElement { Expression: { } spreadExpression } spreadElement] }
                 && spreadElement.IteratorBody is BoundExpressionStatement expressionStatement)
             {
-                var spreadElementHasIdentityConversion = expressionStatement.Expression is not BoundConversion;
+                var spreadElementConversion = expressionStatement.Expression is BoundConversion { Conversion: var actualConversion } ? actualConversion : Conversion.Identity;
+                var spreadElementHasCompatibleConversion = spreadElementConversion.Kind is ConversionKind.Identity or ConversionKind.ImplicitReference;
                 var spreadTypeOriginalDefinition = spreadExpression.Type!.OriginalDefinition;
 
-                if (spreadElementHasIdentityConversion
+                if (spreadElementHasCompatibleConversion
                     && tryGetToArrayMethod(spreadTypeOriginalDefinition, WellKnownType.System_Collections_Generic_List_T, WellKnownMember.System_Collections_Generic_List_T__ToArray, out MethodSymbol? listToArrayMethod))
                 {
                     var rewrittenSpreadExpression = VisitExpression(spreadExpression);
-                    return _factory.Call(rewrittenSpreadExpression, listToArrayMethod.AsMember((NamedTypeSymbol)spreadExpression.Type!));
+                    var listToArray = _factory.Call(rewrittenSpreadExpression, listToArrayMethod.AsMember((NamedTypeSymbol)spreadExpression.Type!));
+                    return _factory.Convert(arrayType, listToArray, spreadElementConversion);
```

Comment:
Somehow I find the usage of `spreadElementConversion` a bit wonky here, since it is based on the element conversion, but is being used here to convert the containing array.

However I think this code itself is correct. If the spread element conversion is `T` to `T` (identity), then `T[]` to `T[]` will be identity, and if it's `TDerived` to `TBase` (implicit reference) then `TDerived[]` to `TBase[]` will be implicit reference.

Reply:
Resolved since `spreadElementConversion` is no longer used here